### PR TITLE
OPENEUROPA-207: Module scaffolding

### PR DESCRIPTION
**Pull request by sandervd**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+```
+$ docker-compose up -d
+```
+
+Then:
+
+```
+$ docker-compose exec -u web web composer install
+$ docker-compose exec -u web web ./vendor/bin/run drupal:site-setup
+$ docker-compose exec -u web web ./vendor/bin/run drupal:site-install
```
Comment (pfrenssen) [inline README.md]: Works great! :+1:

File: README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+```
+
+Your test site will be available at [http://localhost:8080/build](http://localhost:8080/build).
+
+To run the behat test:
```
Comment (pfrenssen) [inline README.md]: This implies there is a Behat test but there is none supplied.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+```
+$ docker-compose exec -u web web ./vendor/bin/behat
+```
```
Comment (pfrenssen) [inline README.md]: Information is missing on how to run PHPUnit tests, this is more important than Behat tests in scope of a module.

File: runner.yml.dist [hunk 4]
```diff
@@ -0,0 +1,28 @@
+drupal:
+  root: "build"
+  base_url: "http://web:8080/build"
+  database:
+    host: "mysql"
+    port: "3306"
+    name: "drupal"
```
Comment (pfrenssen) [inline runner.yml.dist]: Replace `"drupal"` with `"module-skeleton"` so the module name will be used as the database name. Using a fixed database name across multiple different module repositories increases the risk that a developer will accidentally lose their data when switching from one module to another.

File: module-skeleton.info.yml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+name: module-skeleton
```
Comment (pfrenssen) [inline module-skeleton.info.yml]: This should be the human readable name, not the machine name of the module. We can ask the user for this information.

File: replace-names.sh [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+repo_root=$(pwd);
+repo_name=$(basename ${repo_root});
+subst="s/module-skeleton/${repo_name}/g";
+find ${repo_root} -maxdepth 1 -type f | xargs sed -i  ${subst};
+mv module-skeleton.info.yml ${repo_name}.info.yml
+sed -i "/replace-names/d" composer.json
+rm replace-names.sh
```
Comment (pfrenssen) [inline replace-names.sh]: A developer's bash-fu might be weak. Let's add 3 lines of documentation here:

```
# Replace all occurrences of `module-skeleton` with the module name.
...
# Rename the info file.
...
# Remove the reference to this script from the composer manifest and delete it.
```

**Comment by pfrenssen** [comment]
The file `grumphp.yml.dist` is missing. This becomes evident when creating a new project based on the template and testing it on Travis CI. It then spouts errors like the following:

```
PHP Warning:  array_key_exists() expects parameter 2 to be array, null given in /var/www/html/vendor/phpro/grumphp/src/Configuration/Compiler/TaskCompilerPass.php on line 37
PHP Stack trace:
PHP   1. {main}() /var/www/html/vendor/phpro/grumphp/bin/grumphp:0
PHP   2. GrumPHP\Console\Application->__construct() /var/www/html/vendor/phpro/grumphp/bin/grumphp:30
PHP   3. GrumPHP\Console\Application->getContainer() /var/www/html/vendor/phpro/grumphp/src/Console/Application.php:52
```

For instructions on how to build the `grumphp.yml.dist` file, check https://github.com/ec-europa/oe-code-review/blob/master/README.md

File: composer.json [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        "composer/installers": "^1.2",
+        "drupal/console": "^1",
+        "drush/drush": "^9",
+        "drupal/core": "^8.5",
+        "drupal/drupal-extension": "^3.3"
+    },
+    "scripts": {
+        "drupal-scaffold": "DrupalComposer\\DrupalScaffold\\Plugin::scaffold",
+        "post-install-cmd": [
+            "@drupal-scaffold"
```
Comment (pfrenssen) [inline composer.json]: The `@drupal-scaffold` entries should all be removed. This is causing the slow process of scaffolding the Drupal core files to be unnecessarily performed every single time a composer command is executed.

According to the documentation of drupal-scaffold the first line that contains the class should be enough. Ref. https://github.com/drupal-composer/drupal-scaffold/blob/master/README.md

File: composer.json [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        ]
+    },
+    "extra": {
+        "installer-paths": {
+            "build/core": ["type:drupal-core"],
+            "build/modules/contrib/{$name}": ["type:drupal-module"],
+            "build/themes/contrib/{$name}": ["type:drupal-theme"]
+        }
+    }
+}
```
Comment (pfrenssen) [inline composer.json]: We are missing the reference to the Composer facade on drupal.org. We need this to be able to include Drupal modules in our dependency list:

```
    "repositories": [
        {
            "type": "composer",
            "url": "https://packages.drupal.org/8"
        }
    ]
```

File: composer.json [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    "require-dev": {
+        "cweagans/composer-patches": "~1.0",
+        "ec-europa/oe-code-review": "^0.1",
+        "ec-europa/oe-task-runner": "^0.3",
+        "drupal-composer/drupal-scaffold": "^2.2",
+        "composer/installers": "^1.2",
+        "drupal/console": "^1",
+        "drush/drush": "^9",
+        "drupal/core": "^8.5",
+        "drupal/drupal-extension": "^3.3"
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply composer.json]: Also: for `drupal/drupal-extension`, update the version string to `^4.0.0@alpha`. The 3.x branch of DrupalExtension has some serious shortcomings that impact D8 development.

File: README.md (hunk 2 above)
Comment (voidtek) [reply README.md]: I added the behat test on travis.

File: README.md (hunk 3 above)
Comment (voidtek) [reply README.md]: I added the phpunit test too.

File: composer.json (hunk 8 above)
Comment (voidtek) [reply composer.json]: Added

File: composer.json (hunk 7 above)
Comment (voidtek) [reply composer.json]: I removed all @drupal-scaffold entries.

**Pull request closed without merging**